macro-app-profile (#957)
spectrocloud/librarium: content/docs/04.5-devx/05-app-profile.md

## App Profiles

App Profiles are templates created with preconfigured services required for sandbox cluster deployment. App Profiles provide a way to drive consistency across sandbox clusters. You can create as many profiles as required. An App Profile can be created to meet specific types of Workload on Palette Sandbox Clusters. For example, you may create a Development App Profile with Helm Charts, Manifest, different database applications, etc. 

> <br />

Reviewer: Lets mention about two sets of parameters:

1. Input parameters - What user can specify and will be consumed by the current tier. It can even use output variables from below tier.
2. Output parameters - which will be generated once the tier gets deployed. It can be consumed by the above tiers.

> 	                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  

Reviewer: Mention that in the code editor type {{.spectro.}} to see the list available variables in that particular tier.

Author: ![image](https://user-images.githubusercontent.com/49595451/199450815-8f03f47a-66b7-4dc1-bc3d-6b2449022d9a.png)

kind: Secret
metadata:
  name: mysecret
  namespace: {{.spectro.system.apptier.NAMESPACE}}
data:
  

>   username: "{{.spectro.system.apptier.NAMESPACE}}" # Resolves to a value of the variable defined in App Profile tier parameters.yaml file.

Reviewer 2: We can remove username example

>   password: "{{.spectro.app.$appDepName-mongodb.PASSWORD}}" # To refer the tier output variables of the top tiers.

Reviewer 3: We should avoid adding comments in our code snippets as people often copy and paste them. We can move the information provided in the comment to above the code snippet.

Author: done

>   password: "{{.spectro.app.$appDepName-mongodb.PASSWORD}}" # To refer the tier output variables of the top tiers.

Reviewer: To refer to output parameters of the below tiers

</WarningBox>

# App Profile Macros

> Pack macros are the variables defined in the App Profile, and these variables are resolved only at the cluster deployment time.

Reviewer 3: ```suggestion
Pack macros are the variables defined in the App Profile and are only resolved at the cluster deployment time.
```
I simplified the statement a bit.

> <br />

Reviewer 2: Replace `Pack` with `App`

```
{{.spectro.system.<variable.name>}}

> ```

Reviewer 2: Replace `<variable.name>` with `[VARIABLE_NAME]`

**Supported Variables**
* appprofile.name
* appprofile.uid
* appdeployment.name
* appdeployment.uid
* appdeployment.tiername : Resolves to a string value in the format `<deployment name>-<tier name>`

> 

Reviewer 3: We might want to consider a table here. If the last variable has a definition, then the rest should also have a definition. That's why a table is helpful in this context, as it will be easier for people to see the variables and their definitions.

</WarningBox>

# App Profile Macros

Pack macros are the variables defined in the App Profile and are only resolved at the cluster deployment time. The app profile variables can be:

* directly given by the developers while modelling the application profiles.

* resolved during the app instatiation.

> * funnelled/inherited from the lower tiers of the profile.

Reviewer 3: ```suggestion
* inherited from the lower tiers of the App Profile.
```

The variables are of two types:

* [The Input Parameters](/devx/app-profile#inputparameters)

* [The Output Parameters](/devx/app-profile#outputparameters)
<br />

## Input parameters

> What user can specify and will be consumed by the current tier. It can even use output variables from below tier. This parameter resolves to a value of the variable defined in App Profile tier parameters.yaml file.

Reviewer 3: ```suggestion
User-provided values that will be consumed by the current tier. It can even use output variables from below tier. This parameter resolves to a value of the variable defined in App Profile tier parameters.yaml file.
```
@resmaa what does this mean `It can even use output variables from below tier`. It's clear what this means.
I also made the sentence more active vs passive by stating, "User-provided values vs what a user can specify". We always want to use direct voice when possible.

Author: yes

```
{{.spectro.system.[VARIABLE_NAME]}}
```
|**Supported Variables**|  
|-----------------------|
|appprofile.name| 
|appprofile.uid| 
|appdeployment.name| 
|appdeployment.uid|

> |appdeployment.tiername: Resolves to a string value in the format `<deployment name>-<tier name>`.|

Reviewer 3: Make this a two columns table with the description in the other columns. Also, if one row has a description, then all other rows should also have a description.

<br />

The parameter which will be generated once the tier gets deployed. It can be consumed by the above tiers. It refers to the tier output variables of the top tiers. Refer the format and example below:

<br />

```
{{.spectro.system.apptier.<tierInput_Variable_Name>}}
```

**Example**
{{.spectro.system.apptier.NAMESPACE}}: 

## Output parameters

> The variables which are generated once the tier gets deployed. It can be consumed by the higher tiers.

Reviewer 3: ```suggestion
The variables are generated once the tier gets deployed. Output variables can be consumed by the higher tiers in the App Profile.
```
Let's be specific of what can consume the output variables.
